bftengine/tests/testViewChange/test_views_manager.cpp

```diff
@@ -0,0 +1,208 @@
+        replicaId, initialView, ReplicaAsksToLeaveViewMsg::Reason::ClientRequestTimeout));
+
+    viewsManager->storeComplaint(std::move(complaint));
+  }
+
+  ASSERT_EQ(viewsManager->hasQuorumToLeaveView(), true);
+}
+
+TEST_F(ViewsManagerTest, status_message_with_complaints) {
+  bftEngine::impl::SeqNum lastExecutedSeqNum = 200;
+  const bool viewIsActive = true;
+  const bool hasNewChangeMsg = true;
+  const bool listOfPPInActiveWindow = false, listOfMissingVCMsg = false, listOfMissingPPMsg = false;
+
+  for (int replicaId = 0; replicaId < rc.numReplicas; ++replicaId) {
+    if (replicaId == rc.replicaId) continue;
+
+    std::unique_ptr<ReplicaAsksToLeaveViewMsg> complaint(ReplicaAsksToLeaveViewMsg::create(
+        replicaId, initialView, ReplicaAsksToLeaveViewMsg::Reason::ClientRequestTimeout));
+
+    viewsManager->storeComplaint(std::move(complaint));
+  }
+
+  bftEngine::impl::ReplicaStatusMsg replicaStatusMessage(rc.getreplicaId(),
+                                                         initialView,
+                                                         lastStableSeqNum,
+                                                         lastExecutedSeqNum,
+                                                         viewIsActive,
+                                                         hasNewChangeMsg,
+                                                         listOfPPInActiveWindow,
+                                                         listOfMissingVCMsg,
+                                                         listOfMissingPPMsg);
+
+  viewsManager->addComplaintsToStatusMessage(replicaStatusMessage);
+
+  for (const auto& i : viewsManager->getAllMsgsFromComplainedReplicas()) {
+    ASSERT_TRUE(replicaStatusMessage.hasComplaintFromReplica(i.first));
+  }
+}
+
+TEST_F(ViewsManagerTest, get_quorum_for_next_view_on_view_change_message_with_enough_complaints) {
+  bftEngine::impl::ReplicaId sourceReplicaId = (rc.replicaId + 1) % rc.numReplicas;
+  std::set<bftEngine::impl::ReplicaId> otherReplicas;
+  const int nextView = initialView + 1;
+  const int viewToComplainAbout = initialView;
+
+  ViewChangeMsg viewChangeMsg = ViewChangeMsg(sourceReplicaId, nextView, lastStableSeqNum);
+
+  for (int i = 0; i < rc.numReplicas; ++i) {
+    if (i == rc.replicaId || i == sourceReplicaId) continue;
+    otherReplicas.insert(i);
+  }
+
+  ASSERT_LE(rc.fVal + 1, otherReplicas.size());
+
+  for (int complaintNumber = 0; complaintNumber <= rc.fVal; ++complaintNumber) {
+    viewChangeMsg.addComplaint(ReplicaAsksToLeaveViewMsg::create(
+        *std::next(otherReplicas.begin(), complaintNumber),  // Add F + 1 Different complaints
+        viewToComplainAbout,
+        ReplicaAsksToLeaveViewMsg::Reason::ClientRequestTimeout));
+  }
+
+  ViewChangeMsg::ComplaintsIterator iter(&viewChangeMsg);
+  char* complaint = nullptr;
+  MsgSize size = 0;
+
+  while (!viewsManager->hasQuorumToLeaveView() && iter.getAndGoToNext(complaint, size)) {
```

Comment:
Is it possible to call the method **ViewsManager::processComplaintsFromViewChangeMessage** instead of iterating over the complaints?

Reply:
As this function is on a separate branch, let's first merge this PR, and address the comment on the appropriate branch.